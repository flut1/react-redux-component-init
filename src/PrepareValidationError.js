import ExtendableError from './utils/ExtendableError';

class PrepareValidationError extends ExtendableError {}

export default PrepareValidationError;
