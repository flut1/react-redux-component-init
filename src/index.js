export { default as withInitAction } from './withInitAction';
export { default as initReducer } from './reducer';

export { default as initComponent } from './actions/initComponent';
export { default as prepareComponent } from './actions/prepareComponent';
export { default as prepareComponents } from './actions/prepareComponents';
export { default as setInitMode } from './actions/setInitMode';

export { default as PrepareValidationError } from './PrepareValidationError';
export * from './initMode';
export * from './initSelfMode';
export * as actionTypes from './actions/actionTypes';
