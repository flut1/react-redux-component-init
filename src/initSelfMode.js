/**
 * Possible values for the `selfInit` option of `withInitAction()`
 */

/**
 * The component will render immediately, even if `initAction` is still pending.
 * It is recommended to use this option and render a loading indicator or placeholder content
 * until `initAction` is resolved. This will give the user immediate feedback that something is
 * being loaded. While the `initAction` is pending, an `isInitializing` prop will be passed to
 * the component.
 * @type {string}
 */
export const INIT_SELF_ASYNC = 'ASYNC';

/**
 * This will cause the component not to be mounted until the first initialization has completed.
 * The component will remain mounted during any reinitialization.
 * @type {string}
 */
export const INIT_SELF_BLOCKING = 'BLOCKING';

/**
 * The same behavior as `INIT_SELF_BLOCKING` but it will also unmount the component during
 * reinitialization.
 * @type {string}
 */
export const INIT_SELF_UNMOUNT = 'UNMOUNT';

/**
 * Only execute initAction on the server (initMode == MODE_PREPARE). Initialization will
 * be skipped on the client.
 * @type {string}
 */
export const INIT_SELF_NEVER = 'NEVER';
