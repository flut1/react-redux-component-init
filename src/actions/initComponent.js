import propNameValuesToObject from '../utils/propNameValuesToObject';
import { MODE_INIT_SELF, MODE_PREPARE } from '../initMode';
import { INIT_SELF_NEVER } from '../initSelfMode';

import { INIT_COMPONENT } from './actionTypes';

/**
 * Action that initializes a single Component. Can be used both during a component prepare
 * and a component self-init.
 *
 * If the current `initMode` is `MODE_PREPARE` and the component has not yet initialized
 * when this action is dispatched, an error will be thrown.
 *
 * **There is probably no need to directly use this action outside of this module.** This
 * action is automatically dispatched by the `withInitAction()` HoC and the `prepareComponent()`
 * thunk.
 *
 * @param {react.Component} Component The React component to initialize. Passing a component that
 * was not wrapped with `withInitAction()` will result in an error.
 * @param {Array} initValues An array of values that correspond with the array of `initProps` for
 * this component. These values are either passed through `prepareComponent()` or extracted from
 * the `props` when a component mounts or updates.
 * @param {string} prepareKey A unique identifier for the component and it's `initValues`. This
 * is generated by the `getPrepareKey()` util.
 * @param {object} [options] Additional options
 * @param {boolean} [options.isPrepare = false] True if a component is being prepared, false
 * if a component is being mounted or updated.
 * @returns {function} A thunk function that should be passed directly to the Redux `dispatch`
 * function.
 */
export default (
  Component,
  initValues,
  prepareKey,
  { isPrepare = false } = {},
) => (dispatch, getState) => {
  if (!Component.initConfig) {
    throw new Error('No init config found on Component passed to initComponent');
  }

  const {
    componentId,
    initProps,
    initAction,
    options: { onError, getInitState, initSelf, allowLazy },
  } = Component.initConfig;

  const initState = getInitState(getState());
  if (!initState) {
    throw new ReferenceError('Could not find init state. Did you attach the init reducer?');
  }
  const { mode, prepared } = initState;

  let shouldInitialize = (
    ((mode === MODE_INIT_SELF) && (initSelf !== INIT_SELF_NEVER)) ||
    (isPrepare && (typeof prepared[prepareKey] === 'undefined'))
  );
  if ((mode === MODE_PREPARE) && !isPrepare && allowLazy && !prepared[prepareKey]) {
    shouldInitialize = true;
  }

  if (shouldInitialize) {
    const initPropsObj = propNameValuesToObject(initProps, initValues);

    dispatch({
      type: INIT_COMPONENT,
      payload: {
        complete: false,
        isPrepare,
        prepareKey,
      },
    });

    return Promise.resolve()
      .then(() => {
        const initActionReturn = initAction(initPropsObj, dispatch, getState);

        if (typeof initActionReturn.then !== 'function') {
          const error = new Error(`Expected initAction to return a Promise. Returned an ${typeof initActionReturn} instead. Check the initAction for "${componentId}"`);
          error.isInvalidReturnError = true;
          throw error;
        }

        return initActionReturn;
      })
      .catch((e) => {
        if (onError && !e.isInvalidReturnError) {
          onError(e);
        } else {
          throw e;
        }
      })
      .then((result) => {
        dispatch({
          type: INIT_COMPONENT,
          payload: {
            complete: true,
            isPrepare,
            prepareKey,
          },
        });

        return result;
      });
  } else if (mode === MODE_PREPARE && !isPrepare && !allowLazy) {
    if (typeof prepared[prepareKey] === 'undefined') {
      const initPropsObj = propNameValuesToObject(initProps, initValues);
      throw new Error(`Expected component "${componentId}" to be prepared but prepareComponent has not been called with props: \n${JSON.stringify(initPropsObj)}`);
    } else if (prepared[prepareKey] === false) {
      throw new Error(`Expected component "${componentId}" to be prepared but preparation is still pending`);
    }
  }

  return Promise.resolve();
};
