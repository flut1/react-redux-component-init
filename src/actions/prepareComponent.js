import { MODE_PREPARE } from '../initMode';
import extractValuesForProps from '../utils/extractValuesForProps';

import initComponent from './initComponent';

/**
 * Action that prepares a single component. Duplicate calls with exactly the same
 * `Component` and `props` will be ignored.
 *
 * @param {react.Component} Component The component to prepare. If the component was not
 * wrapped in `withInitAction`, this action will have no effect and returns a Promise that
 * resolves immediately.
 * @param {object} [props={}] The values of `initProps` to initialize this component with. These
 * values should be the same as the props that the component will eventually mount with. Omitting
 * one of the `initProps` values configured in `withInitAction()` will result in an error.
 * @returns {function} A thunk function that should be passed directly to the Redux `dispatch`
 * function.
 * @example dispatch(prepareComponent(Comment, { commentId: 5 }));
 */
export default (Component, props = {}) =>
  (dispatch, getState) => {
    // unwrap other HoC if no initConfig is found
    while (!Component.initConfig && Component.WrappedComponent) {
      Component = Component.WrappedComponent; // eslint-disable-line no-param-reassign
    }

    if (Component.initConfig) {
      const {
        componentId,
        initProps,
        options: { getInitState, getPrepareKey },
      } = Component.initConfig;

      const initState = getInitState(getState());
      if (!initState) {
        throw new ReferenceError('Could not find init state. Did you attach the init reducer?');
      }

      const { mode } = initState;
      if (mode === MODE_PREPARE) {
        const initValues = extractValuesForProps(props, initProps);

        initValues.forEach((initValue, index) => {
          if (typeof initValue === 'undefined') {
            throw new ReferenceError(`Component "${componentId}" expected prop "${initProps[index]}" but was not passed to prepareComponent`);
          }
        });

        const prepareKey = getPrepareKey(componentId, initValues);

        return dispatch(initComponent(Component, initValues, prepareKey, {
          isPrepare: true,
        }));
      }
    }

    return Promise.resolve();
  };
