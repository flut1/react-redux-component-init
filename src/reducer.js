import { createSelector } from 'reselect';

import createShallowArrayCompareSelector from './utils/createShallowArrayCompareSelector';
import { INIT_COMPONENT, SET_INIT_MODE } from './actions/actionTypes';
import extractValuesForProps from './utils/extractValuesForProps';
import { MODE_PREPARE } from './initMode';

/**
 * Reducer function that manages the state for `react-redux-component-init`. This reducer should
 * be included using Redux's `combineReducers()` under the `init` key in the root of
 * the store state.
 */
export default (state = {
  mode: MODE_PREPARE,
  prepared: {},
  selfInit: {},
}, action) => {
  switch (action.type) {
    case SET_INIT_MODE:
      return {
        ...state,
        mode: action.payload,
      };
    case INIT_COMPONENT:
      if (action.payload.isPrepare) {
        return {
          ...state,
          prepared: {
            ...state.prepared,
            [action.payload.prepareKey]: action.payload.complete,
          },
        };
      }
      return {
        ...state,
        selfInit: {
          ...state.selfInit,
          [action.payload.prepareKey]: action.payload.complete,
        },
      };
    default:
      return state;
  }
};

const createComponentInitValuesSelector = ({
  componentId, initProps, options: { getPrepareKey },
}) =>
  createShallowArrayCompareSelector(
    (state, props) => extractValuesForProps(props, initProps),
    initValues => ({
      prepareKey: getPrepareKey(componentId, initValues),
      initValues,
    }),
  );

export const createComponentInitStateSelector = initConfig => createSelector(
  createComponentInitValuesSelector(initConfig),
  state => state.selfInit,
  state => state.prepared,
  ({ prepareKey, initValues }, selfInit, prepared) => ({
    prepareKey,
    initValues,
    selfInitState: selfInit[prepareKey],
    isPrepared: !!prepared[prepareKey],
  }),
);
