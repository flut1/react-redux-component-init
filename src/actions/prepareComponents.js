import prepareComponent from './prepareComponent';

/**
 * Simple helper action for preparing multiple components. All components
 * are prepared with the same props.
 *
 * This helper can be useful for preparing a set of route components that were
 * a result of page route matching.
 *
 * @param {Array<react.Component>} components An array of components to initialize.
 * Components that have not been wrapped in `withInitAction` will be ignored.
 * @param {object} props The values of `initProps` that should be used during
 * initialization.
 * @returns {function} A thunk function that should be passed directly to the Redux
 * `dispatch()` function. When passed to `dispatch()` it will return a Promsie that
 * resolves when all components have finished initialization.
 * @example dispatch(prepareComponents([Header,HomePage,Footer], { route: '/home' }));
 */
export default (components, props) => dispatch =>
  Promise.all(components.map(Component => dispatch(prepareComponent(Component, props))));
