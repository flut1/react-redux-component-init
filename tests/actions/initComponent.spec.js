import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import initComponent from '../../src/actions/initComponent';
import withInitAction, { clearComponentIds } from '../../src/withInitAction';
import { MODE_INIT_SELF, MODE_PREPARE } from '../../src/initMode';
import { INIT_COMPONENT } from '../../src/actions/actionTypes';

const mockStore = configureMockStore([thunk]);

describe('initComponent', () => {
  describe('when initMode === MODE_PREPARE', () => {
    describe('with no isPrepare option given', () => {
      describe('and no prepare state is present', () => {
        describe('with no allowLazy option on the component', () => {
          clearComponentIds();
          const store = mockStore({ init: { mode: MODE_PREPARE, prepared: {} } });
          const MockComponent = () => <noscript />;
          const MockWithInit = withInitAction(() => Promise.resolve())(MockComponent);

          it('throws an error', () => {
            expect(() => store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]'))).toThrow();
          });
        });
        describe('with { allowLazy: true } on the component', () => {
          it('does not throw an error', () => {
            clearComponentIds();
            const store = mockStore({ init: { mode: MODE_PREPARE, prepared: {} } });
            const MockComponent = () => <noscript />;
            const MockWithInit = withInitAction(
              () => Promise.resolve(),
              { allowLazy: true },
            )(MockComponent);

            expect(() => store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]'))).not.toThrow();
          });

          it('calls the initAction and dispatches { completed: true } when the action resolves', () => {
            clearComponentIds();
            const store = mockStore({ init: { mode: MODE_PREPARE, prepared: {} } });
            const MockComponent = () => <noscript />;
            const MockWithInit = withInitAction(
              () => Promise.resolve(),
              { allowLazy: true },
            )(MockComponent);

            return store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]')).then(() => {
              const actions = store.getActions();
              expect(actions).toEqual([
                {
                  type: INIT_COMPONENT,
                  payload: { complete: false, isPrepare: false, prepareKey: 'MockComponent[]' },
                },
                {
                  type: INIT_COMPONENT,
                  payload: { complete: true, isPrepare: false, prepareKey: 'MockComponent[]' },
                },
              ]);
            });
          });
        });
      });
      describe('and the preparation has completed', () => {
        describe('with no allowLazy option on the component', () => {
          it('does not throw an error', () => {
            clearComponentIds();
            const store = mockStore({ init: {
              mode: MODE_PREPARE,
              prepared: {
                'MockComponent[]': true,
              },
            } });
            const MockComponent = () => <noscript />;
            const MockWithInit = withInitAction(() => Promise.resolve())(MockComponent);

            expect(() => store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]'))).not.toThrow();
          });

          it('does not dispatch INIT_COMPONENT', () => {
            clearComponentIds();
            const store = mockStore({ init: {
              mode: MODE_PREPARE,
              prepared: {
                'MockComponent[]': true,
              },
            } });
            const MockComponent = () => <noscript />;
            const MockWithInit = withInitAction(() => Promise.resolve())(MockComponent);

            return store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]')).then(() => {
              const actions = store.getActions();
              return expect(actions).toEqual([]);
            });
          });
        });

        describe('with { allowLazy: true } on the component', () => {
          it('does not throw an error', () => {
            clearComponentIds();
            const store = mockStore({ init: {
              mode: MODE_PREPARE,
              prepared: {
                'MockComponent[]': true,
              },
            } });
            const MockComponent = () => <noscript />;
            const MockWithInit = withInitAction(
              () => Promise.resolve(),
              { allowLazy: true },
            )(MockComponent);

            expect(() => store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]'))).not.toThrow();
          });

          it('does not dispatch INIT_COMPONENT', () => {
            clearComponentIds();
            const store = mockStore({ init: {
              mode: MODE_PREPARE,
              prepared: {
                'MockComponent[]': true,
              },
            } });
            const MockComponent = () => <noscript />;
            const MockWithInit = withInitAction(
              () => Promise.resolve(),
              { allowLazy: true },
            )(MockComponent);

            return store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]')).then(() => {
              const actions = store.getActions();
              return expect(actions).toEqual([]);
            });
          });

          it('does not call the initAction', () => {
            clearComponentIds();
            const store = mockStore({ init: {
              mode: MODE_PREPARE,
              prepared: {
                'MockComponent[]': true,
              },
            } });
            const MockComponent = () => <noscript />;
            const mockInitAction = jest.fn(
              () => new Promise(resolve => setTimeout(() => resolve('bar'), 40)),
            );
            const MockWithInit = withInitAction(
              mockInitAction,
              { allowLazy: true },
            )(MockComponent);

            return store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]')).then(
              () => expect(mockInitAction.mock.calls.length).toBe(0),
            );
          });
        });
      });
      describe('and the preparation in state is still pending', () => {
        clearComponentIds();
        const store = mockStore({ init: {
          mode: MODE_PREPARE,
          prepared: {
            'MockComponent[]': false,
          },
        } });
        const MockComponent = () => <noscript />;
        const MockWithInit = withInitAction(() => Promise.resolve())(MockComponent);

        it('throws an error', () => {
          expect(() => store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]'))).toThrow();
        });
      });
      describe('and the preparation is done with different props', () => {
        clearComponentIds();
        const store = mockStore({ init: {
          mode: MODE_PREPARE,
          prepared: {
            'MockComponent["foo"]': true,
          },
        } });
        const MockComponent = () => <noscript />;
        const MockWithInit = withInitAction(['p1'], () => Promise.resolve())(MockComponent);

        it('throws an error', () => {
          expect(
            () => store.dispatch(initComponent(MockWithInit, ['bar'], 'MockComponent["bar"]')),
          ).toThrow();
        });
      });
    });
    describe('with {isPrepare: true}', () => {
      describe('and no prepare state is present', () => {
        describe('with no allowLazy option on the component', () => {
          it('dispatches INIT_COMPONENT with { completed: false }', () => {
            clearComponentIds();
            const store = mockStore({ init: {
              mode: MODE_PREPARE,
              prepared: {},
            } });
            const MockComponent = () => <noscript />;
            const MockWithInit = withInitAction(
              () => new Promise(resolve => setTimeout(() => resolve('bar'), 40)),
            )(MockComponent);
            store.dispatch(initComponent(
              MockWithInit,
              [],
              'MockComponent[]',
              { isPrepare: true },
            ));

            const actions = store.getActions();
            expect(actions).toEqual([
              {
                type: INIT_COMPONENT,
                payload: { complete: false, isPrepare: true, prepareKey: 'MockComponent[]' },
              },
            ]);
          });

          it('calls the initAction and dispatches { completed: true } when the action resolves', () => {
            clearComponentIds();
            const store = mockStore({ init: {
              mode: MODE_PREPARE,
              prepared: {},
            } });
            const MockComponent = () => <noscript />;
            const mockInitAction = jest.fn(
              () => new Promise(resolve => setTimeout(() => resolve('bar'), 40)),
            );
            const MockWithInit = withInitAction(mockInitAction)(MockComponent);
            const initPromise = store.dispatch(initComponent(
              MockWithInit,
              [],
              'MockComponent[]',
              { isPrepare: true },
            ));

            return initPromise.then(() => {
              const actions = store.getActions();
              expect(actions).toEqual([
                {
                  type: INIT_COMPONENT,
                  payload: { complete: false, isPrepare: true, prepareKey: 'MockComponent[]' },
                },
                {
                  type: INIT_COMPONENT,
                  payload: { complete: true, isPrepare: true, prepareKey: 'MockComponent[]' },
                },
              ]);

              expect(mockInitAction.mock.calls.length).toBe(1);
            });
          });
        });
        describe('with { allowLazy: true } on the component', () => {
          it('dispatches INIT_COMPONENT with { completed: false }', () => {
            clearComponentIds();
            const store = mockStore({ init: {
              mode: MODE_PREPARE,
              prepared: {},
            } });
            const MockComponent = () => <noscript />;
            const MockWithInit = withInitAction(
              () => new Promise(resolve => setTimeout(() => resolve('bar'), 40)),
              { allowLazy: true },
            )(MockComponent);
            store.dispatch(initComponent(
              MockWithInit,
              [],
              'MockComponent[]',
              { isPrepare: true },
            ));

            const actions = store.getActions();
            expect(actions).toEqual([
              {
                type: INIT_COMPONENT,
                payload: { complete: false, isPrepare: true, prepareKey: 'MockComponent[]' },
              },
            ]);
          });
        });
      });
      describe('with an initAction that rejects', () => {
        clearComponentIds();
        const store = mockStore({ init: {
          mode: MODE_PREPARE,
          prepared: {},
        } });
        const MockComponent = () => <noscript />;
        const MockWithInit = withInitAction(
          () => new Promise((resolve, reject) => setTimeout(() => reject('bar'), 40)),
        )(MockComponent);

        it(
          'rejects the return promise',
          () => expect(store.dispatch(initComponent(
            MockWithInit,
            [],
            'MockComponent[]',
            { isPrepare: true },
          ))).rejects.toBe('bar'),
        );
      });
      describe('with an initAction that rejects and an error handler', () => {
        clearComponentIds();
        const store = mockStore({ init: {
          mode: MODE_PREPARE,
          prepared: {},
        } });
        const MockComponent = () => <noscript />;
        const mockErrorHandler = jest.fn();
        const MockWithInit = withInitAction(
          () => new Promise((resolve, reject) => setTimeout(() => reject('bar'), 40)),
          { onError: mockErrorHandler },
        )(MockComponent);
        const initPromise = store.dispatch(initComponent(
          MockWithInit,
          [],
          'MockComponent[]',
          { isPrepare: true },
        ));

        it('resolves the returned promise', () => expect(initPromise).resolves.toBeUndefined());
        it(
          'calls the error handler with the rejected value',
          () => expect(mockErrorHandler).toBeCalledWith('bar'),
        );
      });
      describe('and prepare is already in progress', () => {
        clearComponentIds();
        const store = mockStore({ init: {
          mode: MODE_PREPARE,
          prepared: {
            'MockComponent[]': false,
          },
        } });
        const MockComponent = () => <noscript />;
        const mockInitAction = jest.fn(
          () => new Promise(resolve => setTimeout(() => resolve('bar'), 40)),
        );
        const MockWithInit = withInitAction(
          [],
          mockInitAction,
        )(MockComponent);
        const initPromise = store.dispatch(initComponent(
          MockWithInit,
          [],
          'MockComponent[]',
          { isPrepare: true },
        ));

        it('does not call the initAction', () => initPromise.then(
          () => expect(mockInitAction.mock.calls.length).toBe(0)),
        );

        it('does not dispatch INIT_COMPONENT', () => initPromise.then(() => {
          const actions = store.getActions();
          expect(actions).toEqual([]);
        }));
      });
    });
  });
  describe('when initMode === MODE_INIT_SELF', () => {
    describe('with an async initAction', () => {
      it('dispatches INIT_COMPONENT with { completed: false }', () => {
        clearComponentIds();
        const store = mockStore({ init: {
          mode: MODE_INIT_SELF,
          prepared: {},
          selfInit: {},
        } });
        const MockComponent = () => <noscript />;
        const MockWithInit = withInitAction(
          () => new Promise(resolve => setTimeout(() => resolve('bar'), 40)),
        )(MockComponent);
        store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]'));

        const actions = store.getActions();
        expect(actions).toEqual([
          {
            type: INIT_COMPONENT,
            payload: { complete: false, isPrepare: false, prepareKey: 'MockComponent[]' },
          },
        ]);
      });

      it('dispatches { completed: true } when the action resolves', () => {
        clearComponentIds();
        const store = mockStore({ init: {
          mode: MODE_INIT_SELF,
          prepared: {},
          selfInit: {},
        } });
        const MockComponent = () => <noscript />;
        const MockWithInit = withInitAction(
          () => new Promise(resolve => setTimeout(() => resolve('bar'), 40)),
        )(MockComponent);
        const initPromise = store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]'));

        return initPromise.then(() => {
          const actions = store.getActions();
          expect(actions).toEqual([
            {
              type: INIT_COMPONENT,
              payload: { complete: false, isPrepare: false, prepareKey: 'MockComponent[]' },
            },
            {
              type: INIT_COMPONENT,
              payload: { complete: true, isPrepare: false, prepareKey: 'MockComponent[]' },
            },
          ]);
        });
      });

      it('dispatches { completed: true } when the action resolves', () => {
        clearComponentIds();
        const store = mockStore({ init: {
          mode: MODE_INIT_SELF,
          prepared: {},
          selfInit: {},
        } });
        const MockComponent = () => <noscript />;
        const MockWithInit = withInitAction(
          () => new Promise(resolve => setTimeout(() => resolve('bar'), 40)),
        )(MockComponent);
        const initPromise = store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]'));

        return initPromise.then(result => expect(result).toBe('bar'));
      });
    });
  });
  describe('on a component with custom getInitState', () => {
    clearComponentIds();
    const store = mockStore({ foo: { mode: MODE_PREPARE, prepared: { 'MockComponent[]': true } } });
    const MockComponent = () => <noscript />;
    const MockWithInit = withInitAction(() => Promise.resolve(), {
      getInitState: state => state.foo,
    })(MockComponent);

    it('runs without error', () => {
      expect(
        () => store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]')),
      ).not.toThrow();
    });
  });
  describe('with an initAction that does not return a Promise', () => {
    clearComponentIds();
    const store = mockStore({ init: { mode: MODE_INIT_SELF, prepared: {} } });
    const MockComponent = () => <noscript />;
    const MockWithInit = withInitAction(
      () => 5,
    )(MockComponent);

    it('rejects the returned promise', () => {
      expect(
        store.dispatch(initComponent(MockWithInit, [], 'MockComponent[]')),
      ).rejects.toBeDefined();
    });
  });
});
