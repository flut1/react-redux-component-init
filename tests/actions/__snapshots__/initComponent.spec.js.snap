// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initComponent when initMode === MODE_INIT_SELF with an async initAction dispatches { complete: true } when the action resolves 1`] = `
Array [
  Object {
    "payload": Object {
      "complete": false,
      "isPrepare": false,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
  Object {
    "payload": Object {
      "complete": true,
      "isPrepare": false,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
]
`;

exports[`initComponent when initMode === MODE_INIT_SELF with an async initAction dispatches INIT_COMPONENT with { complete: false } 1`] = `
Array [
  Object {
    "payload": Object {
      "complete": false,
      "isPrepare": false,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
]
`;

exports[`initComponent when initMode === MODE_PREPARE with { caller: 'didMount' }option given, no prepare state present and { allowLazy: true } calls the initAction and dispatches { complete: true } when the action resolves 1`] = `
Array [
  Object {
    "payload": Object {
      "complete": false,
      "isPrepare": false,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
  Object {
    "payload": Object {
      "complete": true,
      "isPrepare": false,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
]
`;

exports[`initComponent when initMode === MODE_PREPARE with {caller: 'prepareComponent'} and no prepare state is present with { allowLazy: true } on the component dispatches INIT_COMPONENT with { complete: false } 1`] = `
Array [
  Object {
    "payload": Object {
      "complete": false,
      "isPrepare": true,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
]
`;

exports[`initComponent when initMode === MODE_PREPARE with {caller: 'prepareComponent'} and no prepare state is present with no allowLazy option on the component calls the initAction and dispatches { complete: true } when the action resolves 1`] = `
Array [
  Object {
    "payload": Object {
      "complete": false,
      "isPrepare": true,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
  Object {
    "payload": Object {
      "complete": true,
      "isPrepare": true,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
]
`;

exports[`initComponent when initMode === MODE_PREPARE with {caller: 'prepareComponent'} and no prepare state is present with no allowLazy option on the component dispatches INIT_COMPONENT with { complete: false } 1`] = `
Array [
  Object {
    "payload": Object {
      "complete": false,
      "isPrepare": true,
      "prepareKey": "SimpleInitTestComponent[]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
]
`;
